packages/dds/sequence/src/intervalCollection.ts

```diff
@@ -1156,6 +1161,200 @@ class EndpointIndex<TInterval extends ISerializableInterval> implements Interval
+		private readonly client: Client,
+	) {
+		this.intervalTree = new RedBlackTree<TInterval, TInterval>((a: TInterval, b: TInterval) => {
+			const compareEndsResult = helpers.compareEnds(a, b);
+			if (compareEndsResult !== 0) {
+				return compareEndsResult;
+			}
+			/**
+			 * The Symbol is only applied on the transient interval, which is always intended to be the first input
+			 * in the comparison. Therefore, it is sufficient to only check the Symbol of 'a' in the comparison logic.
+			 */
+			if ((a as any)[EndpointInRangeIndex.alwaysCompareSmaller]) {
+				return -1;
+			}
+			if ((a as any)[EndpointInRangeIndex.alwaysCompareLarger]) {
+				return 1;
+			}
+			const aId = a.getIntervalId();
+			const bId = b.getIntervalId();
+			if (aId !== undefined && bId !== undefined) {
+				return aId.localeCompare(bId);
+			}
+			return 0;
+		});
+	}
+
+	public add(interval: TInterval): void {
+		this.intervalTree.put(interval, interval);
+	}
+
+	public remove(interval: TInterval): void {
+		this.intervalTree.remove(interval);
+	}
+
+	public findIntervalsWithEndpointInRange(start: number, end: number) {
+		if (start <= 0 || start > end || this.intervalTree.isEmpty()) {
+			return [];
+		}
+		const results: TInterval[] = [];
+		const action: PropertyAction<TInterval, TInterval> = (node) => {
+			results.push(node.data);
+			return true;
+		};
+
+		const transientStartInterval = this.helpers.create(
+			"transient",
+			start,
+			start,
+			this.client,
+			IntervalType.Transient,
+		);
+		(transientStartInterval as any)[EndpointInRangeIndex.alwaysCompareSmaller] = true;
+
+		const transientEndInterval = this.helpers.create(
+			"transient",
+			end,
+			end,
+			this.client,
+			IntervalType.Transient,
+		);
+		(transientEndInterval as any)[EndpointInRangeIndex.alwaysCompareLarger] = true;
+
+		this.intervalTree.mapRange(action, results, transientStartInterval, transientEndInterval);
+		return results;
+	}
+}
+
+class StartpointInRangeIndex<TInterval extends ISerializableInterval>
+	implements IStartpointInRangeIndex<TInterval>
+{
+	private readonly intervalTree;
+	private static readonly alwaysCompareSmaller = Symbol();
+	private static readonly alwaysCompareLarger = Symbol();
+
+	constructor(
+		private readonly helpers: IIntervalHelpers<TInterval>,
+		private readonly client: Client,
+	) {
+		this.intervalTree = new RedBlackTree<TInterval, TInterval>((a: TInterval, b: TInterval) => {
+			if ("compareStarts" in helpers && typeof helpers.compareStarts === "function") {
+				const compareStartsResult = helpers.compareStarts(a, b);
+				if (compareStartsResult !== 0) {
+					return compareStartsResult;
+				}
+			}
+			/**
+			 * The Symbol is only applied on the transient interval, which is always intended to be the first input
+			 * in the comparison. Therefore, it is sufficient to only check the Symbol of 'a' in the comparison logic.
+			 */
+			if ((a as any)[StartpointInRangeIndex.alwaysCompareSmaller]) {
```

Comment:
you should factor out this bit to share core with the endpoint index. Maybe a `compareOverrideables(a: Partial<HasComparisonOverride>, b: Partial<HasComparisonOverride>)`.